### ISSUE TITLE
Feature request: Client Factories.

### ISSUE THREAD

**Issue by bslatner** [aws/aws-sdk-net#642]
An abstraction I use all over the place in my code is `IAmazonXxxClientFactory` with a single method that creates an AWS client. Example:

```csharp
public interface IAmazonSQSClientFactory {
    IAmazonSQS CreateClient();
}
```

For developing locally, we load credentials from a stored profile. For production code, or Lambdas, we use IAM roles. In some cases, we need to consume services from multiple regions. Our dependency injector picks the appropriate factory based on the environment.

I used to have all these in one place. But, as the new SDK separates each service into its own assembly -- which I'm in favor of -- I have an equal number of factory assemblies to match.

Where this gets tricky is that, as we consume more AWS services, the more factory assemblies we end up with. And, as teams develop libraries that depend on AWS services, we'd like to make sure the same factories are used everywhere.

It seems a simple, helpful thing to add to the SDK assemblies themselves. I can't imagine I'm the only one who faces this challenge.

**Comment by PavelSafronov** [comment]
What sort of an addition did you have in mind?

**Comment by bslatner** [comment]
The simple answer is just to add interfaces to each SDK assembly for each client defined in that assembly. For example:

``` csharp
public interface IAmazonSQSClientFactory {
    IAmazonSQS CreateClient();
}
```

There's no need to even add any functionality. But, I might also go a step further and add this to Core:

```csharp
public interface IAWSCredentialsFactory {
    AWSCredentials CreateCredentials();
}

public class StoredProfileAWSCredentialsFactory : IAWSCredentialsFactory {
    private readonly string _ProfileName;

    public StoredProfileAWSCredentialsFactory(string profileName) {
        _ProfileName = profileName;
    }

    public AWSCredentials CreateCredentials() {
        return new StoredProfileAWSCredentials(_ProfileName); 
    }
}

public class InstanceProfileAWSCredentialsFactory : IAWSCredentialsFactory {
    public AWSCredentials CreateCredentials() {
        return new InstanceProfileAWSCredentials(_ProfileName); 
    }
}

// other factories for environment, explicit AK and SK, etc, go here
```
With those classes and interfaces in hand, I can easily develop client factories:

```csharp
public SQSClientFactory : IAmazonSQSClientFactory {
    private readonly RegionEndpoint _Region;
    private readonly IAWSCredentialsFactory _CredentialsFactory;

    public SQSClientFactory(RegionEndpoint region, IAWSCredentialsFactory credentialsFactory) { 
        _Region = region;
        _CredentialsFactory = credentialsFactory;
    }

    public IAmazonSQS CreateClient() {
        return new AmazonSQSClient(_CredentialsFactory.CreateCredentials(), _Region);
    }
}
```

If you'd like to explore the problem more deeply, the reason I use these factories is because the default config-based method for creating clients is not sufficient to meet my needs. For example:

1. I frequently consume services in multiple regions. Thus, the single "region" attribute in the `<aws>` config element doesn't work for me.

2. In my environment(s), I know for certain whether I am using role-based credentials or profile-based credentials. There's no need for the SDK to figure this out for me. And, in any event, the SDK uses the same method for every client type, which may not be appropriate. For example, if I have a customer with an S3 bucket I need to talk to, I might have to use stored credentials on a per-bucket basis, but I may also have my owns buckets that need role-based credentials (and they may be in different regions).

3. I might have multiple libraries that need different Dynamo DB table prefixes.

The bottom line is, the config file approach works well if you have homogeneous allocation of resources in a single region and you follow a consistent pattern of naming and such. This is obviously the preferred way to do things, but sometimes this approach is impractical.

**Comment by normj** [comment]
Could you give an example of how you would consume the SDK with these interfaces? I'm really curious what the end user experience would be like. Also what would the experience be if the user needed to do advanced configuration with the service clients config object?

**Comment by bslatner** [comment]
Well, I can show you the way that **I** use them. 

Basically, whenever I have a need for an AWS service in a class, I use my DI framework to inject a factory, like so:

```csharp
public class Foo : IFoo {
    private IAmazonSQSFactory _SQSFactory;

    public Foo(IAmazonSQS sqsClientFactory) {
    }

    public void DoThing() {
        var client = _SQSFactory.CreateClient();
        client.Enqueue("message");
    }
}
```

The point is, my classes neither know nor care where the client comes from or how it is constructed.

Then, during DI setup, I do something like this:

```csharp
Kernel.Bind<IFoo>().To<Foo>();
Kernel.Bind<IAmazonSQSFactory>().To<StoredProfileSQSFactory>().WithConstructorArgument("profileName", "theProfileIWantToUse");
```

Again, this only comes into play when I need to be able to construct clients in more than one way. If I'm only dealing with a single profile in a single region, I can just use the config file.
